Ccda author encounter provenance flag

adunsulag opened the pull request:
Added an office contact setting in care coordination.  Added the
participants piece to the header.

Removed the fDate formatting on the document timestamp as PHP was
already sending a timestamp formatted document and the fDate was being
corrupted.  Had to do this to pass ETT validation.

Added a checkbox to let people force the author timestamp value to be
the encounter date time stamp.  This made it possible to override the
timestamps for the document across the three different test scenarios
without having to change it for each individual patient.  Most people
will have this flag turned off, but it sped up the ONC testing instead
of having to set the timestamp on each doc generation (for both b1 and
g9 testing).

@sjpadgett I'm not sure why its trying to merge my first commit again into your branch, not sure if there is an easy way to skip the merge conflicts there.

adunsulag committed:
Fix ccda author,custodian,config settings

Made it so the author npi works and that the author doesn't blow up when
it returns as a string.

Fixed issue with the custodian uuid accidently being set as the phone
number.

Added in the author date time stamp override in the config settings.

adunsulag committed:
Callback participant, force encounter date flag

Added an office contact setting in care coordination.  Added the
participants piece to the header.

Removed the fDate formatting on the document timestamp as PHP was
already sending a timestamp formatted document and the fDate was being
corrupted.  Had to do this to pass ETT validation.

Added a checkbox to let people force the author timestamp value to be
the encounter date time stamp.  This made it possible to override the
timestamps for the document across the three different test scenarios
without having to change it for each individual patient.  Most people
will have this flag turned off, but it sped up the ONC testing instead
of having to set the timestamp on each doc generation (for both b1 and
g9 testing).

adunsulag committed:
Remove extraneous letter in send screen.

adunsulag committed:
Merge branch 'ccda_author' of github.com:sjpadgett/openemr into ccda_author

adunsulag committed:
Merge branch 'ccda_author' into ccda_author_encounter_provenance_flag

adunsulag commented on ccdaservice/serveccda.js:
```diff
@@ -2346,7 +2389,9 @@ function populateHeader(pd) {
-            "date": fDate(pd.created_time_timezone),
+            // php already sends this in YYYYMMDDHHmmss(+/-)zzzz format
+            // so we can just take it straight away for the document head
+            "date": pd.created_time_timezone,
```
@sjpadgett You may want to look at this closely.  I thought this was all working but I must not have reset my node server when I tested with ETT originally.  The fDate formatting was changing the timezone to an invalid format that failed the ETT validation (it wants the timestamps to be without any spaces or hyphens).

adunsulag committed:
Add referring participant refactor participants

Refactored the participants on the php side to be an array so we can add
participants in the future.  Added the referral participant

adunsulag commented:
@sjpadgett I just pushed my changes with the xml format and the participants.  The referring provider is there.  Was having date problems so I just pushed it in and I'll work on fixing the datetime population.

sjpadgett commented on interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/EncounterccdadispatchTable.php:
```diff
@@ -486,6 +487,54 @@ public function getInformant($pid, $encounter)
+
+        // TODO: get this working later
+//        $referralDate = DateFormatterUtils::dateStringToDateTime($referral['refer_date']);
+        $referralDate = date('YmdHisO');
```
Could be date('Y-m-d H:i:sO') then it will get sorted in moment if fDate() isn't used. Seems we have three diff ISO formats where moment bulks at the YmdHisO format. Still fDate will format correctly.

adunsulag replied:
sounds good, I'll give it a try.

sjpadgett commented:
Looks good Stephen. thanks

adunsulag committed:
Fix document effectiveTime, participant datetime

Got the effectiveTime back to using moment like Jerry and I discussed.

Got the participant time working since referral date has its own format
regardless of the system format.

adunsulag commented:
@sjpadgett This should be working now.  I also fixed the document effective time to use the fDate && moment library like we discussed.

adunsulag committed:
Add telecom to participant

Wasn't passing ETT test cases w/o telecom info.

adunsulag committed:
Fix g9 filtering problems

Had problems on the single date filtering test.  This commit fixes
those.

adunsulag committed:
Fix g9 date displays, stylesheet timezone bug

Had a bug where the timezone was showing but no hour/minute because they
get hidden if the time is 00:00.  Made the timezone hidden if the
hour/minute is also hidden.

Fixed the g9 service dates not rendering because they were populated in
a range request.

adunsulag committed:
Fixed zip and street address for participants.

Got the referring and contact address into the stylesheet.  Definitely
not ideal as the header is huge.  Will have to touch up the styles and
add in the fallback PCP if the REFerral isn't there for our human
readable view.

adunsulag commented:
@sjpadgett I've got the referring provider and office stuff in now.  Fixed a number of issues I found with the participants and hopefully everything is now working.  

I believe what's outstanding is the timezone changes on the entry / sub components for author and the encompassing encounter.  Thanks for all your work here. 

I did add a bunch of g9 stuff intermixed here so hopefully that's ok my stuff is dependent on your other changes here so I just put it all together.  Feel free to bring it in whenever you'd like.

sjpadgett commented:
@adunsulag A really nice job on jumping in on this CCDA stuff. Not an easy thing plus, now you own it!:) teehee
